merino-cache/src/memory.rs

```diff
@@ -379,15 +382,24 @@ mod tests {
         assert!(!cache.contains_key(&"expired".to_owned()));
 
         // Verify the reported metric.
-        assert_eq!(rx.len(), 2);
+        assert_eq!(rx.len(), 3);
         let collected_data: Vec<String> = rx
             .iter()
```

Comment:
You've hit a footgun w/ this channel `Receiver` from `SpyMetrics` -- it's a ["blocking" iterator](https://docs.rs/crossbeam-channel/latest/crossbeam_channel/struct.Receiver.html#method.iter), where "_Each call to [next](https://doc.rust-lang.org/nightly/core/iter/traits/iterator/trait.Iterator.html#tymethod.next) blocks waiting for the next message and then returns it._".

The `take(2)` call issued to the iterator previously avoided its blocking (and we proved there was only 2 messages to take by asserting so above). You've removed that call, so the iteration continues instead of stopping -- `collect` gets the 3rd message but issues another `next` call. Instead of signaling the Iterator's exhausted (returning `None` to `next`) it blocks, waiting for the other side of the channel to send a new value which never happens. Hence the test timing out in CI.

Issuing `take(3)` fixes this but there's a better fix I've been meaning to apply everywhere:

```suggestion
            .try_iter()
```

This uses a [non-blocking `Iterator`](https://docs.rs/crossbeam-channel/latest/crossbeam_channel/struct.Receiver.html#method.try_iter) instead so we don't have to limit the iteration via `take`. 

Can you switch to `try_iter` here and additionally fix `merino-suggest-providers/src/providers/keyword_filter.rs` (its `test_provider_all_filtered` also relies on `take()`, let's get rid of it)

Reply:
Fascinating @pjenvey ! @ncloudioj was incredibly generous with his time today to give me an excellent overview of iterators, adapters and their properties so I learned a ton and this adds some interesting extra nuances of detail.  That very clearly explains the blocking/non-blocking nature of the two methods which I will be sure to make note of!

The removal of `take()` originally was intended to remove a redundant method, given we had just above checked that there were 3 dummy elements in that vector. However, upon your explanation, I better understand why it was there. I will make those adjustments now.

Reply:
Indeed, I have no idea how it passed on my machine @ncloudioj .Maybe the M1 architecture is just so good, it fixes compiler errors on the fly 😆 Either way, thank you @pjenvey  for catching that. Glad it resolves something you just found out about!